lib/client.js

```diff
@@ -0,0 +1,169 @@
+var WS = require('ws');
+var debug = require('debug')('docker-exec-websocket-server:lib:client');
+var url = require('url');
+var events = require('events');
+var _ = require('lodash');
+var assert = require('assert');
+var msgcode = require('../lib/messagecodes.js');
+var through = require('through');
+
+export default class DockerExecWebsocketClient extends events.EventEmitter {
+  constructor (options) {
+    super();
+    this.options = _.defaults({}, options, {
+      tty: true,
+      command: ['/bin/bash'],
+      protocol: 'ws',
+      slashes: true,
+    });
+  }
+
+  /* Makes a client program with unbroken stdin, stdout, stderr streams
+   * Is also an EventEmitter with 'exit' event
+   *
+   * Required options:
+   * url parts (hostname port pathname)
+   * or url
+   * tty: whether or not we expect VT100 style output
+   * command: array or string of command to be run in exec
+   * not supported yet: ssl cert
+   */
+  async execute () {
+    this.options.query = {
+      tty: encodeURIComponent(this.options.tty ? 'true' : 'false'),
+      command: this.options.command,
+    };
+    this.url = this.options.url || url.format(this.options);
```

Comment:
Let's just say that we should never specify `options.url` because then options like `tty` and `command` won't be added...

IMO we could require an input url and then be explicit like this:
```
let url = this.options.url + '?' + querystring.stringify({
  tty: this.options.tty ? 'true' : 'false',
  command: this.options.command
})
```

---
Btw, I didn't know this, but:
```js
var querystring = require('querystring');
var s = querystring.stringify({tty: 'true', command: ['a', 'b']});
// renders s = 'tty=true&command=a&command=b'
```

And `querystring.parse` will reverse this... So we don't need JSON encoding of `command`, but I suppose you already figured that out :)